### PR TITLE
[Refactor] refactor losses function, support key `reduction_override`

### PR DESCRIPTION
**Pull request by BIGWangYuDong**
Firstly, I used `weight_reduce_loss` to rewrite the forward function and, added `reduction` key in `__init__`, the reduction defaults to mean. Meanwhile, I refactored the GHMC, GHMR forward function in `ghm_loss.py`, which make the loss function can support the key `reduction_override `.

Then I supplemented the losses test unit, to ensure that the losses of the same task can be replaced.

I have confirmed the results of GHMC and GHMR are the same as before the modification.

### CODE REVIEW

**Comment by ZwwWayne** [comment]
We can consider add a unit test that uses Faster R-CNN as a base config, then change to use different regression loss in the config and run forward. This could ensure compatibility in the interfaces of each loss.

File: tests/test_models/test_loss_compatibility.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    }
+    img_metas = [img_matas]
+    cfg_model = ConfigDict(cfg_model)
+
+    # verify bbox loss function compatibility
+    for loss_bbox in loss_bboxes:
+        cfg_model.roi_head.bbox_head.loss_bbox = loss_bbox
+        model = build_detector(cfg_model)
+        loss = model.forward_train(imgs, img_metas, gt_bboxes, gt_labels)
+        assert isinstance(loss['loss_cls'], torch.Tensor)
```
Comment (ZwwWayne) [inline tests/test_models/test_loss_compatibility.py:184]: we usually test whether loss is greater than 0

File: tests/test_models/test_loss_compatibility.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+
+def test_loss_compatibility():
+    """Test loss_cls and loss_bbox compatibility.
+
+    Using Faster R-CNN as a sample, modifying the loss function in the config
+    file to verify the compatibility of Loss APIS
+    """
+    # Faster R-CNN config dict
+    cfg_model = dict(
```
Comment (ZwwWayne) [inline tests/test_models/test_loss_compatibility.py:15]: Can we directly use _base_/models/faster_rcnn_r50_fpn.py?

**Comment by BIGWangYuDong** [comment]
We split the `test_loss_compatibility` unit into `test_cls_loss_compatibility` and `test_box_loss_compatibility` 

We first use `_demo_mm_inputs` from test_forward.py to load the config file and build the detector. Then, we use `_get_detector_cfg` to generate a  non-empty truth batch and feed it into the `detector.forward_train`. It is important to set roi_head.bbox_head.reg_decoded_bbox be True when using IoULoss.

We rewrite the test unit to support pytest and change the way of assertion. In the latest version, we assert whether loss is greater than 0.

File: tests/test_models/test_loss_compatibility.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
+import pytest
+from test_forward import _demo_mm_inputs, _get_detector_cfg
```
Comment (ZwwWayne) [inline tests/test_models/test_loss_compatibility.py:2]: may put this into the function to avoid modifing setup_cfg

File: mmdet/models/losses/ghm_loss.py [hunk 4]
```diff
@@ -3,6 +3,7 @@
 import torch.nn.functional as F
 
 from ..builder import LOSSES
+from .utils import weight_reduce_loss
```
Comment (ZwwWayne) [inline mmdet/models/losses/ghm_loss.py:6]: need to rebenchmark a retinanet-GHM

**Comment by BIGWangYuDong** [comment]
Being re-benchmark retinanet-GHM on the slurm, will get the result tomorrow

**Comment by BIGWangYuDong** [comment]
Being re-benchmark retinanet-GHM on the slurm, will get the result tomorrow

**Comment by BIGWangYuDong** [comment]
@ZwwWayne 
got 37.4 or 37.1 mAP on R-50-FPN, which is higher than the original 37.0 mAP.
image

**Review by ZwwWayne — APPROVED** [review]
